### ISSUE TITLE
TFLite model weight shape is different from original model

### ISSUE THREAD

**Issue by tu1258** [tensorflow/tensorflow#44215]
I use TF2.3.0 and try two different way. One is python API and the other is command line `tflite_convert`.
**Python API**
```
import os
import sys
import absl.logging as logging
import numpy as np
import tensorflow.compat.v1 as tf

np.set_printoptions(threshold=sys.maxsize)
tf.reset_default_graph()
config = tf.ConfigProto()
config.gpu_options.allow_growth = True
sess = tf.Session(config=config)
tf.keras.backend.set_session(sess)

converter = tf.lite.TFLiteConverter.from_saved_model('best_model')
converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS]
converter.allow_custom_ops = True
quant_best_model = converter.convert()
with open('quant_best_model.tflite', 'wb') as w:
  w.write(quant_best_model)
```

**Command line**
```
tflite_convert \
--output_file 'tflite_convert.tflite' \
--saved_model_dir 'best_model'
```

[best_weights.zip](https://github.com/tensorflow/tensorflow/files/5418846/best_weights.zip)
[best_model.zip](https://github.com/tensorflow/tensorflow/files/5418507/best_model.zip)

Open `best_weights` and `quant_best_model.tflite` or `tflite_convert.tflite` in [netron](https://lutzroeder.github.io/netron/) and inspect the weight shape. Actually `quant_best_model.tflite` and `tflite_convert.tflite` should be same.

last conv2D layer in tflite model (the dtype is float32 because the tensor is too small to be quantized, no problem here)
![image](https://user-images.githubusercontent.com/59950011/96791211-9c655280-142a-11eb-9113-a1baa0e9ea3d.png)

last conv2D layer in original model
![image](https://user-images.githubusercontent.com/59950011/96791262-b2731300-142a-11eb-93da-fa315d56eb3c.png)

Actually it's the transpose version if you look into how weights distributed.

The tflite model just fail like the model that generate random output in my case.
Any advice? Thanks!

**Comment by tu1258** [comment]
@ravikyram @ymodak Sorry for pushing you guys, but the weekend is coming and I know that there won't be any response for a few days. Please comment if there is any advice.

**Comment by karimnosseir** [comment]
@tu1258 Is your question about why the shape is transposed ? That's because TFLite expects the weights to be OHWI, and we transpose the values during conversion from TF to TFLite, so we don't have to do this during inference.

As for "The tflite model just fail like the model that generate random output in my case."
I am not sure i understand this statement.
Do you mean you tested a model and the resulting values is different between TF and TFLite ?
If yes, then please include the model and a reproduce code snippet.
Note that, if you are quantizing the model, then loss in accuracy is expected.

Thanks

**Comment by tu1258** [comment]
@karimnosseir thanks for the reply.

> tested a model and the resulting values is different between TF and TFLite
yes. And the accuracy is between 8~9 for a 12 class classification task, which seems like just random output. The original model accuracy is around 95.

**Comment by karimnosseir** [comment]
"please include the model and a reproduce code snippet.
Note that, if you are quantizing the model, then loss in accuracy is expected."

**Comment by tu1258** [comment]
The code above should be enough for generate tflite model.

**Comment by tu1258** [comment]
@karimnosseir 

Code (include both contain model and generating final output)
```
import os
import sys
import absl.logging as logging
import numpy as np
import tensorflow.compat.v1 as tf

np.set_printoptions(threshold=sys.maxsize)
tf.reset_default_graph()
config = tf.ConfigProto()
config.gpu_options.allow_growth = True
sess = tf.Session(config=config)
tf.keras.backend.set_session(sess)

read_fingerprints = np.loadtxt('./testbench/data.csv', delimiter=',')

test_fingerprints = read_fingerprints.reshape(-1,16384).astype(np.float32)
print('type(test_fingerprints):', type(test_fingerprints))
print('shape(test_fingerprints):', test_fingerprints.shape)

train_dir = 'tcn/kws_7x36_1'

converter = tf.lite.TFLiteConverter.from_saved_model(train_dir + '/best_model')
converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS]
converter.allow_custom_ops = True
quant_best_model = converter.convert()
with open(train_dir + '/quant_best_model.tflite', 'wb') as w:
  w.write(quant_best_model)


interpreter = tf.lite.Interpreter(model_path=os.path.join(train_dir + '/quant_best_model.tflite'))
interpreter.allocate_tensors()
input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()
inputs = []
for s in range(len(input_details)):
  inputs.append(np.zeros(input_details[s]['shape'], dtype=np.float32))
  
model = tf.keras.models.load_model(train_dir + '/best_model', custom_objects={'tf': tf})

read_fingerprints = np.loadtxt('./testbench/data.csv', delimiter=',')
test_fingerprints = read_fingerprints.reshape(1,16384)
model.run_eagerly = True
print('tf.executing_eagerly:',tf.executing_eagerly())
print('model.run_eagerly:',model.run_eagerly)

interpreter.set_tensor(input_details[0]['index'], test_fingerprints.astype(np.float32))
interpreter.invoke()
out_tflite = interpreter.get_tensor(output_details[0]['index'])
print('out_tflite:',out_tflite)


intermediate_layer_model = tf.keras.Model(inputs=model.input, outputs=model.get_layer(index=len(model.layers)-1).output)
intermediate_output = intermediate_layer_model(test_fingerprints)
print('intermediate_output:',intermediate_output)
```

The final output I generate
```
out_tflite: [[12.354025   -1.9960475  -0.01582483 -0.13261858 -2.7005167  -1.4998152
  -3.1499515  -3.5536532  -3.20407    -0.65802205 -4.0220547   2.5022275 ]]
intermediate_output: tf.Tensor(
[[12.102602    0.08953991 -3.884885   -0.07776461 -4.7590914  -0.7605376
  -3.9976068  -4.774014   -3.9251728  -1.4000554  -5.652582    2.8998525 ]], shape=(1, 12), dtype=float32)
```

**Comment by karimnosseir** [comment]
@srjoglekar246 Can you have a look on the accuracy issue ?

Thanks

**Comment by srjoglekar246** [comment]
@tu1258 To help me debug this faster, could you clarify: If you convert to TFLite without quantization/optimization, does the output look okay? (i.e. without `converter.optimizations = [tf.lite.Optimize.DEFAULT]`) If yes, this might be a bug with our quantization tooling and I can point this to them. If the float model is also wrong, then the issue is likely with the converter.
